banyand/property/repair.go

```diff
@@ -0,0 +1,853 @@
+	nodes := make([]*repairTreeNode, 0)
+	if parent.tp == repairTreeNodeTypeRoot {
+		// reading the slot nodes
+		if err = r.seekPosition(r.footer.leafNodeFinishedOffset, io.SeekStart); err != nil {
+			return nil, fmt.Errorf("seeking to slot node offset %d in file %s failure: %w", r.footer.leafNodeFinishedOffset, r.file.Name(), err)
+		}
+		slotDataBytes := make([]byte, r.footer.slotNodeLen)
+		if _, err = io.ReadFull(r.reader, slotDataBytes); err != nil {
+			return nil, fmt.Errorf("reading slot node data from file %s failure: %w", r.file.Name(), err)
+		}
+		reduceBytesLen := int64(len(slotDataBytes))
+
+		var slotNodeIndex, leafStartOff, leafLen int64
+		var slotShaVal []byte
+		for reduceBytesLen > 0 {
+			slotDataBytes, slotNodeIndex, err = encoding.BytesToVarInt64(slotDataBytes)
+			if err != nil {
+				return nil, fmt.Errorf("decoding slot node index from file %s failure: %w", r.file.Name(), err)
+			}
+			slotDataBytes, slotShaVal, err = encoding.DecodeBytes(slotDataBytes)
+			if err != nil {
+				return nil, fmt.Errorf("decoding slot node sha value from file %s failure: %w", r.file.Name(), err)
+			}
+			slotDataBytes, leafStartOff, err = encoding.BytesToVarInt64(slotDataBytes)
+			if err != nil {
+				return nil, fmt.Errorf("decoding slot node leaf start offset from file %s failure: %w", r.file.Name(), err)
+			}
+			slotDataBytes, leafLen, err = encoding.BytesToVarInt64(slotDataBytes)
+			if err != nil {
+				return nil, fmt.Errorf("decoding slot node leaf length from file %s failure: %w", r.file.Name(), err)
+			}
+			reduceBytesLen = int64(len(slotDataBytes))
+			nodes = append(nodes, &repairTreeNode{
+				shaValue:  fmt.Sprintf("%x", slotShaVal),
+				id:        fmt.Sprintf("%d", slotNodeIndex),
+				tp:        repairTreeNodeTypeSlot,
+				leafStart: leafStartOff,
+				leafLen:   leafLen,
+			})
+		}
+		return nodes, nil
+	} else if parent.tp == repairTreeNodeTypeLeaf {
+		return nil, nil
+	}
+
+	// otherwise, reading the leaf nodes
+	err = r.seekPosition(parent.leafStart, io.SeekStart)
+	if err != nil {
+		return nil, fmt.Errorf("seeking to leaf node offset %d in file %s failure: %w", r.footer.leafNodeFinishedOffset, r.file.Name(), err)
+	}
+	leafDataBytes := make([]byte, parent.leafLen)
+	if _, err = io.ReadFull(r.reader, leafDataBytes); err != nil {
+		return nil, fmt.Errorf("reading leaf node data from file %s failure: %w", r.file.Name(), err)
+	}
+	reduceBytesLen := int64(len(leafDataBytes))
+	for reduceBytesLen > 0 {
+		var entity, shaVal []byte
+		leafDataBytes, entity, err = encoding.DecodeBytes(leafDataBytes)
+		if err != nil {
+			return nil, fmt.Errorf("decoding leaf node entity from file %s failure: %w", r.file.Name(), err)
+		}
+		leafDataBytes, shaVal, err = encoding.DecodeBytes(leafDataBytes)
+		if err != nil {
+			return nil, fmt.Errorf("decoding leaf node sha value from file %s failure: %w", r.file.Name(), err)
+		}
+		reduceBytesLen = int64(len(leafDataBytes))
+		nodes = append(nodes, &repairTreeNode{
```

Comment:
You should use the streaming method to load leaves. Loading all data into memory at once will cause OOM errors. Providing a readLine method for retrieving leaves is a better solution.